src/btree/bt_cursor.c

```diff
@@ -847,10 +847,9 @@ __wt_btcur_search(WT_CURSOR_BTREE *cbt)
     if (session->format_private != NULL && (ret == 0 || ret == WT_NOTFOUND))
         session->format_private(ret, session->format_private_arg);
 
-#ifdef HAVE_DIAGNOSTIC
-    if (ret == 0)
-        WT_ERR(__wt_cursor_key_order_init(cbt));
-#endif
+    if (DIAGNOSTIC_ASSERTS_ENABLED(session))
```

Comment:
Is there a reason we chose to not merge these if's? i.e.
```
    if (DIAGNOSTIC_ASSERTS_ENABLED(session) && ret == 0)
            WT_ERR(__wt_cursor_key_order_init(cbt));
```
Similarly for __btcur_search_neighboring.

Reply:
No specific reason but was an oversight. Thanks! [4ec1860](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/4ec1860657e6aeef19c7f7a20fa12817c7c791a1)

- [ ]

Reply:
@ajmorton you made a node that the DIAGNOSTIC_ASSERTS check pulls a pointer off the session, is it better to check the `ret` first?

Reply:
Interesting point..
The `ASSERT_ENABLED` is wrapped inside an `UNLIKELY` under the hood so branch prediction will always assume it's false. My gut feel is that the order of args won't make a big difference at runtime.

To add to that we've scaled back the use of our runtime flag from pretty much every `WT_ASSERT` in the codebase to only a subset of locations like this one, so the impacts of pointer derefs will be much less than the 10% perf drop we saw previously. This also doesn't look like a critical code path (the actual searcing will take much longer) so we can won't need to worry too much.

Finally on a stylistic note I like having `ASSERT_ENABLED` as the first arg. To me it indicates a context switch from normal code flow to diagnostic mode which is more relevant to the reader than the `ret` check. I'd even suggest keeping these two conditions as separate `if`s for readability.

Reply:
I am happy to leave it as is then I'd prefer not to split it back into two if statements. One thing worth calling out is that transforming `#ifdef HAVE_DIAGNOSTIC` into a `DIAGNOSTIC_ASSERTS_ENABLED` check is a bit odd when the what the diagnostic code was doing was setting it up to check something at a later date, or even just wrapping some expensive code. If you understand the context that by default `DIAGNOSTIC_ASSERTS_ENABLED` is true for a `DIAGNOSTIC` build then it is clear what it is doing but it requires extra context than the original `#ifdef HAVE_DIAGNOSTIC` block. Not sure what we could do here as adding a code comment for every usage of it seems excessive.

Reply:
👍  
WT-10428 created to update the `DIAGNOSTIC_ASSERTS_ENABLED` to something better.